File: shared/engine/src/commonTest/kotlin/com/pastura/engine/RandomSourceTests.kt
package com.pastura.engine

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertNotEquals
import kotlin.test.assertTrue

/**
 * Pins the ADR-023 S3b RNG seam to known-answer vectors shared with the Swift
 * `Pastura/PasturaTests/Engine/RandomSourceTests.swift`. Every vector below is
 * the same literal on both sides: a change to either engine that breaks one is
 * a parity break, whatever the single-language tests say.
 */
class RandomSourceTests {

    @Test
    fun splitMix64MatchesReferenceVectorsForSeedZero() {
        val source = SplitMix64RandomSource(seed = 0uL)
        assertEquals(
            listOf(0xE220A8397B1DCDAFuL, 0x6E789E6AA1B965F4uL, 0x06C45D188009454FuL, 0xF88BB8A8724C81ECuL),
            List(4) { source.nextUInt64() },
        )
    }

    @Test
    fun splitMix64MatchesReferenceVectorsForNonZeroSeed() {
        val source = SplitMix64RandomSource(seed = 0x123456789ABCDEF0uL)
        assertEquals(
            listOf(0x161922C645CE50E8uL, 0xAD760CAFA1697B60uL, 0x3501FF44902CA50DuL, 0x417CB9A826D831DFuL),
            List(4) { source.nextUInt64() },
        )
    }

    /**
     * [index] is `% n` and nothing cleverer — the Swift twin pins the same four
     * indices for the same seed.
     */
    @Test
    fun indexReducesByModulo() {
        val source = SplitMix64RandomSource(seed = 0uL)
        assertEquals(listOf(2, 1, 2, 4), List(4) { source.index(below = 7) })
    }

    /** [unit] takes the top 53 bits, so the doubles are exact and shared. */
    @Test
    fun unitTakesTopFiftyThreeBits() {
        val source = SplitMix64RandomSource(seed = 0uL)
        assertEquals(
            listOf(0.8833108082136426, 0.43152799704850997, 0.026433771592597743, 0.9708819781538285),
            List(4) { source.unit() },
        )
    }

    /** The same seed restarts the same stream; a different seed does not. */
    @Test
    fun seedDeterminesTheStream() {
        val first = SplitMix64RandomSource(seed = 42uL)
        val same = SplitMix64RandomSource(seed = 42uL)
        val other = SplitMix64RandomSource(seed = 43uL)
        val stream = List(8) { first.nextUInt64() }
        assertEquals(stream, List(8) { same.nextUInt64() })
        assertNotEquals(stream, List(8) { other.nextUInt64() })
    }

    /** An empty pool is a handler bug, not a runtime condition to absorb. */
    @Test
    fun indexRejectsAnEmptyPool() {
        assertFailsWith<IllegalArgumentException> {
            SplitMix64RandomSource(seed = 0uL).index(below = 0)
        }
    }

    /**
     * The production source draws from the platform generator: helper values
     * stay in range and the raw stream is not constant.
     */
    @Test
    fun systemSourceHelpersStayInRange() {
        val source = SystemRandomSource()
        repeat(64) {
            assertTrue(source.index(below = 5) in 0..4)
            val unit = source.unit()
            assertTrue(unit >= 0.0 && unit < 1.0)
        }
        assertTrue(List(16) { source.nextUInt64() }.toSet().size > 1)
    }
}

File: Pastura/PasturaTests/Engine/RandomSourceTests.swift
import Foundation
import Testing

@testable import Pastura

/// Pins the ADR-023 S3b RNG seam to known-answer vectors shared with
/// `shared/engine/src/commonTest/.../RandomSourceTests.kt`. A change to either
/// side that breaks a vector here is a parity break, whatever the Swift-only
/// tests say.
struct RandomSourceTests {
  /// Reference SplitMix64 output for seed 0 (matches Vigna's reference
  /// implementation) and for an arbitrary non-zero seed.
  @Test(arguments: [
    (
      UInt64(0),
      [0xE220_A839_7B1D_CDAF, 0x6E78_9E6A_A1B9_65F4, 0x06C4_5D18_8009_454F, 0xF88B_B8A8_724C_81EC]
    ),
    (
      UInt64(0x1234_5678_9ABC_DEF0),
      [0x1619_22C6_45CE_50E8, 0xAD76_0CAF_A169_7B60, 0x3501_FF44_902C_A50D, 0x417C_B9A8_26D8_31DF]
    )
  ])
  func splitMix64MatchesReferenceVectors(seed: UInt64, expected: [UInt64]) {
    let source = SplitMix64RandomSource(seed: seed)
    #expect((0..<4).map { _ in source.nextUInt64() } == expected)
  }

  /// `index(below:)` is `% n` and nothing cleverer — the Kotlin twin pins the
  /// same four indices for the same seed.
  @Test func indexReducesByModulo() {
    let source = SplitMix64RandomSource(seed: 0)
    #expect((0..<4).map { _ in source.index(below: 7) } == [2, 1, 2, 4])
  }

  /// `unit()` takes the top 53 bits, so the doubles are exact and shared.
  @Test func unitTakesTopFiftyThreeBits() {
    let source = SplitMix64RandomSource(seed: 0)
    let expected = [
      0.8833108082136426, 0.43152799704850997, 0.026433771592597743, 0.9708819781538285
    ]
    #expect((0..<4).map { _ in source.unit() } == expected)
  }

  /// The same seed restarts the same stream; a different seed does not.
  @Test func seedDeterminesTheStream() {
    let first = SplitMix64RandomSource(seed: 42)
    let same = SplitMix64RandomSource(seed: 42)
    let other = SplitMix64RandomSource(seed: 43)
    let stream = (0..<8).map { _ in first.nextUInt64() }
    #expect(stream == (0..<8).map { _ in same.nextUInt64() })
    #expect(stream != (0..<8).map { _ in other.nextUInt64() })
  }

  /// The production source draws from the system generator: helper values
  /// stay in range and the raw stream is not constant.
  @Test func systemSourceHelpersStayInRange() {
    let source = SystemRandomSource()
    for _ in 0..<64 {
      #expect((0..<5).contains(source.index(below: 5)))
      let unit = source.unit()
      #expect(unit >= 0.0 && unit < 1.0)
    }
    #expect(Set((0..<16).map { _ in source.nextUInt64() }).count > 1)
  }
}
